File: Kalorie/Kalorie/Features/MyCreatedMeal/MyCreatedMealEditorViewModel.swift
//
//  MyCreatedMealEditorViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 20.08.2026.
//

import Foundation

struct MyCreatedMealIngredientDraft: Identifiable {

    // MARK: - Properties

    let id = UUID()
    let item: FoodItemDomain
    var gramsText: String
}

final class MyCreatedMealEditorViewModel: ObservableObject {

    // MARK: - Properties

    @Published private(set) var state: LoadingState<Void> = .idle
    @Published var name: String
    @Published var ingredients: [MyCreatedMealIngredientDraft]
    @Published var searchText = ""
    @Published private(set) var searchResults: [FoodItemDomain] = []
    @Published private(set) var externalSearchResults: [FoodItemDomain] = []
    @Published private(set) var isExternalSearchLoading = false
    @Published var isScannerVisible = false
    @Published var lastScannedBarcode = ""
    @Published private(set) var isBarcodeSearchLoading = false
    @Published private(set) var scannedIngredientId: UUID?
    @Published var alertItem: AlertItem?
    @Published var isSaveConfirmationVisible = false
    @Published private(set) var shouldDismiss = false
    let searchPlaceholder: String

    private let existingMeal: MyCreatedMealDomain?
    private let initialName: String
    private let initialIngredients: [MyCreatedMealIngredientDomain]
    private let searchFoodItems: any SearchFoodItemsUseCaseProtocol
    private let searchFoodExternally: any SearchFoodExternallyUseCaseProtocol
    private let fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    private let fetchFoodByBarcodeExternally: any FetchFoodByBarcodeExternallyUseCaseProtocol
    private let createMyCreatedMeal: any CreateMyCreatedMealUseCaseProtocol
    private let updateMyCreatedMeal: any UpdateMyCreatedMealUseCaseProtocol
    private let onSaved: () -> Void

    var isEditing: Bool { existingMeal != nil }

    var title: String { isEditing ? L10n.MyCreatedMeal.titleEdit : L10n.MyCreatedMeal.titleNew }

    var confirmationTitle: String { isEditing ? L10n.MyCreatedMeal.confirmUpdate : L10n.MyCreatedMeal.confirmCreate }

    var ingredientDomains: [MyCreatedMealIngredientDomain] { Self.ingredientDomains(from: ingredients) }

    var hasChanges: Bool { name != initialName || ingredientDomains != initialIngredients }

    var canSave: Bool {
        guard MyCreatedMealValidation.canSave(name: name, ingredients: ingredientDomains) else { return false }
        return !isEditing || hasChanges
    }

    // MARK: - Init

    init(
        searchFoodItems: any SearchFoodItemsUseCaseProtocol,
        searchFoodExternally: any SearchFoodExternallyUseCaseProtocol,
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol,
        fetchFoodByBarcodeExternally: any FetchFoodByBarcodeExternallyUseCaseProtocol,
        createMyCreatedMeal: any CreateMyCreatedMealUseCaseProtocol,
        updateMyCreatedMeal: any UpdateMyCreatedMealUseCaseProtocol,
        existingMeal: MyCreatedMealDomain? = nil,
        onSaved: @escaping () -> Void = {}
    ) {
        self.existingMeal = existingMeal
        self.searchFoodItems = searchFoodItems
        self.searchFoodExternally = searchFoodExternally
        self.fetchFoodItemByBarcode = fetchFoodItemByBarcode
        self.fetchFoodByBarcodeExternally = fetchFoodByBarcodeExternally
        self.createMyCreatedMeal = createMyCreatedMeal
        self.updateMyCreatedMeal = updateMyCreatedMeal
        self.onSaved = onSaved
        let example = L10n.AddFood.searchExamples.randomElement() ?? ""
        searchPlaceholder = L10n.AddFood.searchPlaceholder(example: example)
        let resolvedName = existingMeal?.name ?? ""
        let resolvedIngredients = (existingMeal?.ingredients ?? []).map { ingredient in
            MyCreatedMealIngredientDraft(
                item: FoodItemDomain(
                    id: ingredient.foodItemId,
                    czName: ingredient.czName,
                    engName: ingredient.engName,
                    weight: ingredient.grams,
                    date: .now,
                    nutrition: ingredient.nutrition
                ),
                gramsText: Self.formattedGrams(ingredient.grams)
            )
        }
        name = resolvedName
        ingredients = resolvedIngredients
        initialName = resolvedName
        initialIngredients = Self.ingredientDomains(from: resolvedIngredients)
    }

    // MARK: - Functions

    @MainActor
    func onSearchTextChanged() async {
        guard !searchText.isEmpty else {
            searchResults = []
            externalSearchResults = []
            return
        }
        do {
            try await Task.sleep(for: .milliseconds(300))
        } catch {
            return
        }
        do {
            searchResults = try await searchFoodItems(query: searchText)
        } catch {
            searchResults = []
        }
        guard searchResults.isEmpty && searchText.count >= 3 else {
            externalSearchResults = []
            return
        }
        isExternalSearchLoading = true
        defer { isExternalSearchLoading = false }
        do {
            externalSearchResults = try await searchFoodExternally(query: searchText)
        } catch {
            externalSearchResults = []
        }
    }

    func onScannerButtonTapped() {
        isScannerVisible = true
    }

    @MainActor
    func onBarcodeScanned() async {
        let barcode = lastScannedBarcode
        guard !barcode.isEmpty else { return }
        lastScannedBarcode = ""
        isBarcodeSearchLoading = true
        defer { isBarcodeSearchLoading = false }
        if let local = try? await fetchFoodItemByBarcode(barcode: barcode) {
            isScannerVisible = false
            scannedIngredientId = onSelectSearchResult(local)
            return
        }
        do {
            if let external = try await fetchFoodByBarcodeExternally(barcode: barcode) {
                isScannerVisible = false
                scannedIngredientId = onSelectSearchResult(external)
                return
            }
        } catch {
            alertItem = AlertItem(title: L10n.AddFood.errorLoadFailed)
            return
        }
        alertItem = AlertItem(title: L10n.AddFood.errorBarcodeNotFound)
    }

    @discardableResult
    func onSelectSearchResult(_ item: FoodItemDomain) -> UUID {
        let draft = MyCreatedMealIngredientDraft(item: item, gramsText: "")
        ingredients.append(draft)
        searchText = ""
        searchResults = []
        externalSearchResults = []
        return draft.id
    }

    func onGramsFieldDefocused(id: UUID) {
        guard
            let index = ingredients.firstIndex(where: { $0.id == id }),
            ingredients[index].gramsText.isEmpty
        else { return }
        ingredients[index].gramsText = "100"
    }

    func onDeleteIngredient(at offsets: IndexSet) {
        ingredients.remove(atOffsets: offsets)
    }

    func onSaveTapped() {
        guard canSave else { return }
        isSaveConfirmationVisible = true
    }

    @MainActor
    func onSaveConfirmed() async {
        guard !state.isLoading else { return }
        state = .loading
        defer { state = .loaded }
        do {
            if let existingMeal {
                try await updateMyCreatedMeal(
                    MyCreatedMealDomain(
                        id: existingMeal.id,
                        name: name,
                        ingredients: ingredientDomains,
                        createdAt: existingMeal.createdAt,
                        updatedAt: existingMeal.updatedAt
                    )
                )
            } else {
                _ = try await createMyCreatedMeal(name: name, ingredients: ingredientDomains)
            }
            onSaved()
            shouldDismiss = true
        } catch {
            alertItem = AlertItem(title: L10n.MyCreatedMeal.errorSaveFailed)
        }
    }

    // MARK: - Private

    private static func ingredientDomains(from ingredients: [MyCreatedMealIngredientDraft]) -> [MyCreatedMealIngredientDomain] {
        ingredients.map { draft in
            MyCreatedMealIngredientDomain(
                foodItemId: draft.item.id,
                czName: draft.item.czName,
                engName: draft.item.engName,
                grams: Self.parsedGrams(draft.gramsText),
                nutrition: draft.item.nutrition
            )
        }
    }

    private static func parsedGrams(_ text: String) -> Double {
        Double(text.replacingOccurrences(of: ",", with: ".")) ?? 0
    }

    private static func formattedGrams(_ value: Double) -> String {
        value.truncatingRemainder(dividingBy: 1) == 0 ? String(Int(value)) : String(value)
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/mycreatedmeal/MyCreatedMealEditorViewModel.kt
package antoni.kalorie.features.mycreatedmeal

import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.components.FoodPortionDraft
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemKind
import antoni.kalorie.core.models.FoodPortionDomain
import antoni.kalorie.core.models.MyCreatedMealDomain
import antoni.kalorie.core.models.MyCreatedMealIngredientDomain
import antoni.kalorie.core.models.MyCreatedMealValidation
import antoni.kalorie.core.usecases.CreateMyCreatedMealUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodByBarcodeExternallyUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodItemsByIdsUseCaseProtocol
import antoni.kalorie.core.usecases.SearchFoodExternallyUseCaseProtocol
import antoni.kalorie.core.usecases.SearchFoodItemsUseCaseProtocol
import antoni.kalorie.core.usecases.UpdateMyCreatedMealUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.core.utils.isLoading
import antoni.kalorie.features.addfoodsheet.AddFoodSheetViewModel
import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

data class MyCreatedMealIngredientDraft(
    val id: UUID = UUID.randomUUID(),
    val item: FoodItemDomain,
    val gramsText: String,
)

class MyCreatedMealEditorViewModel(
    private val searchFoodItems: SearchFoodItemsUseCaseProtocol,
    private val searchFoodExternally: SearchFoodExternallyUseCaseProtocol,
    private val fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCaseProtocol,
    private val fetchFoodByBarcodeExternally: FetchFoodByBarcodeExternallyUseCaseProtocol,
    private val fetchFoodItemsByIds: FetchFoodItemsByIdsUseCaseProtocol,
    private val createMyCreatedMeal: CreateMyCreatedMealUseCaseProtocol,
    private val updateMyCreatedMeal: UpdateMyCreatedMealUseCaseProtocol,
    private val existingMeal: MyCreatedMealDomain? = null,
    private val onSaved: () -> Unit = {},
    private val dismissesOnSave: Boolean = true,
) : ViewModel() {

    // MARK: - Properties

    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    val name = MutableStateFlow(existingMeal?.name ?: "")
    val ingredients = MutableStateFlow(
        (existingMeal?.ingredients ?: emptyList()).map { ingredient ->
            MyCreatedMealIngredientDraft(
                item = FoodItemDomain(
                    id = ingredient.foodItemId,
                    kind = FoodItemKind.CATALOGUE,
                    czName = ingredient.czName,
                    engName = ingredient.engName,
                    weight = ingredient.grams,
                    date = Instant.now(),
                    nutrition = ingredient.nutrition,
                ),
                gramsText = formattedGrams(ingredient.grams),
            )
        },
    )
    val portions = MutableStateFlow(
        (existingMeal?.portions ?: emptyList()).ifEmpty { null }
            ?.map { FoodPortionDraft(name = it.name, gramsText = formattedGrams(it.grams)) }
            ?: listOf(FoodPortionDraft.blank),
    )
    val searchText = MutableStateFlow("")
    private val _searchResults = MutableStateFlow<List<FoodItemDomain>>(emptyList())
    val searchResults: StateFlow<List<FoodItemDomain>> = _searchResults
    private val _externalSearchResults = MutableStateFlow<List<FoodItemDomain>>(emptyList())
    val externalSearchResults: StateFlow<List<FoodItemDomain>> = _externalSearchResults
    private val _isExternalSearchLoading = MutableStateFlow(false)
    val isExternalSearchLoading: StateFlow<Boolean> = _isExternalSearchLoading
    val isScannerVisible = MutableStateFlow(false)
    val lastScannedBarcode = MutableStateFlow("")
    private val _isBarcodeSearchLoading = MutableStateFlow(false)
    val isBarcodeSearchLoading: StateFlow<Boolean> = _isBarcodeSearchLoading
    private val _scannedIngredientId = MutableStateFlow<UUID?>(null)
    val scannedIngredientId: StateFlow<UUID?> = _scannedIngredientId
    val alertItem = MutableStateFlow<AlertItem?>(null)
    val isSaveConfirmationVisible = MutableStateFlow(false)
    private val _shouldDismiss = MutableStateFlow(false)
    val shouldDismiss: StateFlow<Boolean> = _shouldDismiss
    @StringRes val searchExampleRes: Int = AddFoodSheetViewModel.searchExamples.random()

    private val initialName = name.value
    private val initialIngredients = ingredientDomains(ingredients.value)
    private val initialPortions = parsedPortions(portions.value)

    val isEditing: Boolean
        get() = existingMeal != null

    @get:StringRes
    val titleRes: Int
        get() = if (isEditing) R.string.myCreatedMeal_title_edit else R.string.myCreatedMeal_title_new

    @get:StringRes
    val confirmationTitleRes: Int
        get() = if (isEditing) R.string.myCreatedMeal_confirm_update else R.string.myCreatedMeal_confirm_create

    val ingredientDomains: List<MyCreatedMealIngredientDomain>
        get() = ingredientDomains(ingredients.value)

    val portionDomains: List<FoodPortionDomain>
        get() = parsedPortions(portions.value)

    val hasChanges: Boolean
        get() = name.value != initialName || ingredientDomains != initialIngredients || portionDomains != initialPortions

    val canSave: Boolean
        get() {
            if (!MyCreatedMealValidation.canSave(name = name.value, ingredients = ingredientDomains, portions = portionDomains)) return false
            return !isEditing || hasChanges
        }

    // MARK: - Functions

    suspend fun onAppear() {
        if (!isEditing) return
        val freshItems = try {
            fetchFoodItemsByIds(ingredients.value.map { it.item.id })
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.MY_CREATED_MEAL)
            return
        }
        val freshById = freshItems.associateBy { it.id }
        ingredients.value = ingredients.value.map { draft ->
            val fresh = freshById[draft.item.id]
            if (
                fresh != null &&
                (
                    fresh.nutrition != draft.item.nutrition ||
                        fresh.czName != draft.item.czName ||
                        fresh.engName != draft.item.engName
                    )
            ) {
                draft.copy(item = fresh)
            } else {
                draft
            }
        }
    }

    suspend fun onSearchTextChanged() {
        if (searchText.value.isEmpty()) {
            _searchResults.value = emptyList()
            _externalSearchResults.value = emptyList()
            return
        }
        delay(SEARCH_DEBOUNCE_MILLIS)
        try {
            _searchResults.value = searchFoodItems(searchText.value)
        } catch (error: CancellationException) {
            throw error
        } catch (_: Exception) {
            _searchResults.value = emptyList()
        }
        if (_searchResults.value.isNotEmpty() || searchText.value.length < EXTERNAL_SEARCH_MIN_LENGTH) {
            _externalSearchResults.value = emptyList()
            return
        }
        _isExternalSearchLoading.value = true
        try {
            _externalSearchResults.value = searchFoodExternally(searchText.value)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.MY_CREATED_MEAL)
            _externalSearchResults.value = emptyList()
        } finally {
            _isExternalSearchLoading.value = false
        }
    }

    fun onScannerButtonTapped() {
        isScannerVisible.value = true
    }

    suspend fun onBarcodeScanned() {
        val barcode = lastScannedBarcode.value
        if (barcode.isEmpty()) return
        _isBarcodeSearchLoading.value = true
        try {
            try {
                fetchFoodItemByBarcode(barcode)?.let { local ->
                    isScannerVisible.value = false
                    _scannedIngredientId.value = onSelectSearchResult(local)
                    return
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                Log.warning(error, Constants.LogCategory.MY_CREATED_MEAL)
            }
            try {
                fetchFoodByBarcodeExternally(barcode)?.let { external ->
                    isScannerVisible.value = false
                    _scannedIngredientId.value = onSelectSearchResult(external)
                    return
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                Log.error(error, Constants.LogCategory.MY_CREATED_MEAL)
                alertItem.value = AlertItem(titleRes = R.string.addFood_error_loadFailed)
                return
            }
            alertItem.value = AlertItem(titleRes = R.string.addFood_error_barcodeNotFound)
        } finally {
            lastScannedBarcode.value = ""
            _isBarcodeSearchLoading.value = false
        }
    }

    fun onSelectSearchResult(item: FoodItemDomain): UUID {
        val draft = MyCreatedMealIngredientDraft(item = item, gramsText = "")
        ingredients.value = ingredients.value + draft
        searchText.value = ""
        _searchResults.value = emptyList()
        _externalSearchResults.value = emptyList()
        return draft.id
    }

    fun onGramsFieldDefocused(id: UUID) {
        val index = ingredients.value.indexOfFirst { it.id == id }
        if (index < 0 || ingredients.value[index].gramsText.isNotEmpty()) return
        ingredients.value = ingredients.value.toMutableList().also { it[index] = it[index].copy(gramsText = "100") }
    }

    fun onDeleteIngredient(offsets: Set<Int>) {
        ingredients.value = ingredients.value.filterIndexed { index, _ -> index !in offsets }
    }

    fun onSaveTapped() {
        if (!canSave) return
        isSaveConfirmationVisible.value = true
    }

    suspend fun onSaveConfirmed() {
        if (_state.value.isLoading) return
        _state.value = LoadingState.Loading
        try {
            val meal = existingMeal
            if (meal != null) {
                updateMyCreatedMeal(
                    MyCreatedMealDomain(
                        id = meal.id,
                        name = name.value,
                        ingredients = ingredientDomains,
                        createdAt = meal.createdAt,
                        updatedAt = meal.updatedAt,
                        portions = portionDomains,
                    ),
                )
            } else {
                createMyCreatedMeal(name.value, ingredientDomains, portionDomains)
            }
            onSaved()
            _shouldDismiss.value = dismissesOnSave
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MY_CREATED_MEAL)
            alertItem.value = AlertItem(titleRes = R.string.myCreatedMeal_error_saveFailed)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    // MARK: - Private

    private companion object {
        const val SEARCH_DEBOUNCE_MILLIS = 300L
        const val EXTERNAL_SEARCH_MIN_LENGTH = 3

        fun ingredientDomains(ingredients: List<MyCreatedMealIngredientDraft>): List<MyCreatedMealIngredientDomain> =
            ingredients.map { draft ->
                MyCreatedMealIngredientDomain(
                    foodItemId = draft.item.id,
                    czName = draft.item.czName,
                    engName = draft.item.engName,
                    grams = parsedGrams(draft.gramsText),
                    nutrition = draft.item.nutrition,
                )
            }

        fun parsedGrams(text: String): Double = text.replace(',', '.').toDoubleOrNull() ?: 0.0

        fun parsedPortions(drafts: List<FoodPortionDraft>): List<FoodPortionDomain> =
            drafts.mapNotNull { draft ->
                val grams = parsedGrams(draft.gramsText)
                if (grams >= 1) FoodPortionDomain(name = draft.name, grams = grams) else null
            }

        fun formattedGrams(value: Double): String =
            if (value % 1.0 == 0.0) value.toLong().toString() else value.toString()
    }
}
